compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -347,16 +368,29 @@ simplifyExpr world = fst . cata go
         , not (isFreeExprVar x (freeVars (snd e2))) -> e2
 
       -- (let x = e1 in e2).f    ==>    let x = e1 in e2.f
-      -- NOTE(MH): The reason for the choice of `s1` and `s2` is as follows:
-      -- - If `fv(let x = e1 in e2) ⊆ V`, then `fv(e1) ⊆ V` and
-      --   `fv(e2) ⊆ V ∪ {x}`.
-      -- - If `let x = e1 in e2` is k-safe, then `e1` is 0-safe and `e2` is
-      --   k-safe.
-      EStructProjF f (ELet (Binding (x, t) e1) e2, Info fv sf _) ->
+      EStructProjF f (ELet (Binding (x, t) e1) e2, s0) ->
         go $ ELetF (BindingF (x, t) (e1, s1)) (go $ EStructProjF f (e2, s2))
         where
-          s1 = Info fv (sf `min` Safe 0) TCNeither
-          s2 = Info (freeExprVar x <> fv) sf TCNeither
+          (s1, s2) = infoUnstepELet x s0
+
+      -- (λx1 ... xn. e0) e1 ... en    ==>    let x1 = e2 in ... let xn = en in e0,
+      -- if `xi` is not free in `ej` for any `i < j`
+      --
+      -- This rule is achieved by combining the rules for `(λx. e1) e2` and
+      -- `(let x = e1 in e1) e3`, if `x` is not free in `e3`, repeatedly.
+
+      -- (λx. e1) e2    ==>    let x = e2 in e1
+      ETmAppF (ETmLam (x, t) e1, s0) (e2, s2) ->
+        go $ ELetF (BindingF (x, t) (e2, s2)) (e1, s1)
```

Comment:
We don't need to check if `x` is free in `e2` since `let` are _non_-recursive. Would it help to comment to that effect?

I assume the comment you're referring to is the one about repeated application of this rule and the one below. The "if `x` is not free in `e3`" is supposed to resemble the condition of the rule below. Given the confusion it has already caused, I would simply remove the if-clause. This shouldn't make a difference since there's only one rule for `(let x = e1 in e2) e3` anyway.

Reply:
Yeah a comment that lets are not recursive + removing the `if` (it’s already below at the case where there is the condition) sounds good, thanks!